nerfstudio/utils/misc.py

```diff
@@ -155,3 +166,87 @@ def strtobool(val) -> bool:
     FMI https://stackoverflow.com/a/715468
     """
     return val.lower() in ("yes", "y", "true", "t", "on", "1")
+
```

Comment:
most of the changes in this file can be dropped if we move to Python 3.8, right?

I spoke to @tancik and it seems like we can just bump the minimum version, it'll help with #1838 too

Reply:
Dropping support for 3.7 wouldn't help much here. The problem is that __orig_class__ used to be populated before __init__ was called in 3.7, but in 3.8+ it is populated after the __init__ call. We need to access it from the __init__ so I used this workaround.

Reply:
Oops, yeah actually that makes a lot of sense!

To me this seems like what `__class_getitem__()` was designed for: https://docs.python.org/3/reference/datamodel.html#object.__class_getitem__

Incomplete, but as an alternative suggestion what do you think about a pattern that looks like this?


```python
from typing import Any, Generic, TypeVar

from typing_extensions import override

TDataset = TypeVar("TDataset")


class VanillaDataManager(Generic[TDataset]):
    def __init__(self) -> None:
        print(self.get_dataset_type())

    def get_dataset_type(self) -> Any:
        return "the default dataset type"

    def __class_getitem__(cls, key):
        class DataManager(VanillaDataManager):
            @override
            def get_dataset_type(self) -> Any:
                return key

        return DataManager


# Prints "int"
VanillaDataManager[int]()
```

Reply:
I am not as experienced with how type inference is implemented in various tools, but perhaps returning a dynamically defined class may be problematic and less intuitive for users? Otherwise the pattern looks good and much simpler compared to my code.

Reply:
What would happen when you serialise the type with pyyaml (in TrainerConfig). Is it able to serialise and construct a dynamic type?

Reply:
The following code (used in storing and loading configs) does not work with dynamic types:
```
string = yaml.dump(VanillaDataManager[int]())
obj = yaml.load(string, Loader=yaml.UnsafeLoader)
reveal_type(obj)
```

Reply:
Nice catch! Maybe we can use `functools.partial`?

```python
import functools
from typing import Generic, cast

from typing_extensions import TypeVar, reveal_type

TDataset = TypeVar("TDataset", default=str)


class VanillaDataManager(Generic[TDataset]):
    def __init__(self, **kwargs) -> None:
        self.dataset_type = cast(TDataset, kwargs.get("dataset_type", str))
        print(self.dataset_type)

    def __class_getitem__(cls, key):
        return functools.partial(VanillaDataManager, dataset_type=key)


# prints "int"
x = VanillaDataManager[int]()

# type is revealed as Type[int]
reveal_type(x.dataset_type)


# prints "str"
y = VanillaDataManager()

# type is revealed as Type[str]
reveal_type(y.dataset_type)
```

Reply:
The object obtained after loading the yaml will be different from the original (not generic), but I guess it should be fine as long as we do not access __orig_class__ or similar.

Reply:
There is one more problem:
```
class tmp(VanillaDataManager[int]):
    pass
x = tmp()
print(x.dataset_type)
```
The following does not work (since the object returned from __class_getitem__ is not a class I guess?)

Reply:
But I can use this awful thing:
```
    def __class_getitem__(cls, key):
        return type(cls.__name__, (cls,), {"__module__": cls.__module__, "__init__": functools.partialmethod(cls.__init__, dataset_type=key)})
```

Reply:
I have pushed a new version which  uses this.

Reply:
Certainly awful, but it has my support 🙂